hadoop-hdfs-project/hadoop-hdfs/src/site/markdown/NamenodeFGL.md

```diff
@@ -0,0 +1,201 @@
+
+### Global Lock
+
+Used for operations requiring the entire namespace locked.
+
+There are some operations that need to lock the entire namespace, e.g. safe mode related operations, HA service related, etc. NN uses the global lock to make these operations thread-safe. Outside of these infrequent operations that require the global write lock, all other operations have to acquire the global read lock.
+
+### Lock Order
+
+As mentioned above, there are the global lock, DN lock, and full path lock. NN acquires locks in this specific order to avoid deadlocks.
+
+Locks are to be acquired in this order:
+- Global > DN > Full path
+- Global > DN > Last `INodeFile`
+
+Possible lock combinations are as follows:
+- Global write lock
+- Global read lock > Full path lock
+- Global read lock > DN read/write lock
+- Global read lock > DN read/write lock > Read/Write lock of last `INodeFile`
+- Global read lock > DN read/write lock > Full path lock
+
+### Lock Pools
+
+NN allocates locks as needed to the INodes used by active threads, and deletes them after the locks are no longer in use. Locks for commonly accessed `INode`s like the root are cached.
+
+NN uses an `INodeLockPool` to manage these locks. The lock pool:
+- Returns a closeable lock for an INode based on the lock type,
+- Removes this lock if it is no longer used by any threads.
+
+Similar to `INodeLockPool`, a `DNLockPool` is used to manage the locks for DNs. Unlike `INodeLockPool`, `DNLockPool` keeps all locks in memory due to the comparatively lower number of locks.
+
+### Lock Modes
+
+Operations related to namespace tree have different semantics and may involve the modification or access of different INodes, for example: `getBlockLocation` only accesses the last iNodeFile, `delete` modifies both the parent and the last INode, `mkdir` may modify multiple ancestor INodes.
+
+Four lock modes (plus no locking):
+- LOCK_READ
+  - This lock mode acquires the read locks for all INodes down the path.
+  - Example operations: `getBlockLocation`, `getFileInfo`.
+- LOCK_WRITE
+  - This lock mode acquires the write lock for the last INode and the read locks for all ancestor INodes in the full path.
+  - Example operations: `setPermission`, `setReplication`.
+- LOCK_PARENT
+  - This lock mode acquires the write lock for the last two INodes and the read locks for all remaining ancestor INodes in the full path.
+  - Example operations: `rename`, `delete`, `create` (when the parent directory exists).
+- LOCK_ANCESTOR
+  - This lock mode acquires the write lock for the last existing INode and the read locks for all remaining ancestor INodes in the full path.
+  - Example operations: `mkdir`, `create` (when the parent directory doesn't exist).
+- NONE
+  - This lock mode does not acquire any locks for the given path.
+
+Roadmap
+-------
+
+#### Step 1: Split the global lock into FSLock and BMLock
```

Comment:
How about marking these 3 stages?
Currently we nearly finish the 1st stage. can mark it as implemented?(Will merge this doc since finishing all development work)
For 2nd and 3rd stages, mark them Implementing?